h2o-algos/src/main/java/hex/DataInfo.java

```diff
@@ -856,6 +856,30 @@ public void addNum(int id, double val) {
       numVals[i] = val;
     }
 
+    /*
+    This method will perform an inner product of rows.  It will be able to handle categorical data
+    as well as numerical data.  However, the two rows must have exactly the same column types.  This
+    is used in a situation where the rows are coming from the same dataset.
+     */
+    public final double dotSame(Row rowj) {
+      // nums
+      double elementij = 0.0;
+      double weights = this.weight*rowj.weight;
+      for(int i = 0; i < this.nNums; ++i)  {
+        elementij += weights*this.numVals[i]*rowj.numVals[i]; // multiply numerical parts of columns
```

Comment:
minor thing, I'd move the weight multiplication to the very end, to save an op per element